src/components/multi-select/multi-select.jsx

```diff
@@ -0,0 +1,316 @@
+              value={props.optionIDGetter(option)}
+            >
+              {props.optionLabelGetter(option)}
+            </ListOption>
+          ))}
+      </Listbox>
+    );
+  };
+
+  function getOption(id) {
+    return props.options.find(option => props.optionIDGetter(option) === id);
+  }
+
+  function getVisibleOptions() {
+    return props.options
+      .filter(option => props.optionLabelGetter(option).toLowerCase().includes(autocompleteValue.toLowerCase()))
+      .filter(option => !value.some((val) => {
+        const optionID = props.optionIDGetter(option);
+
+        if (typeof optionID === 'string') {
+          return val.toLowerCase() === optionID.toLowerCase();
+        }
+
+        return val === optionID;
+      }));
+  }
+
+  function onAutocompleteBlur() {
+    if (!autocompleteRef.current) return;
+
+    autocompleteRef.current.setCustomValidity('');
+    if (!autocompleteRef.current.validity.valid) {
+      setValidationMessage(autocompleteRef.current.validationMessage);
+    } else {
+      setValidationMessage(props.validationMessage || '');
+    }
+  }
+
+  function onAutocompleteChange(event) {
+    setExpanded(true);
+    setAutocompleteValue(event.target.value);
+  }
+
+  function onDropdownClose() {
+    setExpanded(false);
+    setAutocompleteValue('');
+  }
+
+  function onOptionRemove(event) {
+    const newValue = value.filter((option, index) => (
+      valueRefs[index].current !== event.target
+    ));
+    setValue(newValue);
+  }
+
+  function onOptionsClear(event) {
+    event.preventDefault();
+
+    if (!props.readOnly) {
+      setValue([]);
+      setExpanded(false);
+      autocompleteRef.current.focus();
+    }
+  }
+
+  function onOptionSelect(event) {
+    const selectedOptionIndex = visibleOptionsRefs.findIndex(optionRef => optionRef === event.target);
+    const selectedOption = visibleOptions[selectedOptionIndex];
+    setExpanded(false);
+    const newValue = [...value, props.optionIDGetter(selectedOption)]
+      .sort((a, b) => props.optionIDGetter(a) - props.optionIDGetter(b));
+    setValue(newValue);
+    setAutocompleteValue('');
+    autocompleteRef.current.focus();
+  }
+
+  function onClick(event) {
+    if (event.defaultPrevented) return;
+    if (props.readOnly) return;
+
+    setExpanded(true);
+  }
+
+  function onKeyDown(event) {
+    switch (event.key) {
+      case 'Escape':
+        setExpanded(false);
+        break;
+      default:
+    }
+  }
+
+  useDropdownClose(wrapperRef, expanded, onDropdownClose);
+
+  useEffect(() => {
+    setValue(props.value);
+  }, [props.value.map(val => JSON.stringify(val)).join(',')]);
+
+  useEffect(() => {
+    props.onChange({ target: selfRef.current });
+  }, [value.map(val => JSON.stringify(val)).join(',')]);
+
+  useImperativeHandle(selfRef, () => ({
+    get dirty() {
+      return props.value.map(val => JSON.stringify(val)).join(',') !== this.value.map(val => JSON.stringify(val)).join(',');
```

Comment:
Note: this uses `JSON.stringify` to handle the object option case. It will break if that object contains cyclical references. Are we okay with that? Should we leave a comment about it somewhere?

Reply:
I think that's okay. Most of the time, the values will be IDs anyways. When a developer runs into a cyclical issue then they can just pass down an ID and then use the `onChange` to map the selected ID to their respective object.